components/match2/commons/starcraft_starcraft2/match_group_input_starcraft_ffa.lua

```diff
@@ -183,129 +181,145 @@ function StarcraftFfaInput.MatchWinnerProcessing(match, OppNumber, noscore)
+	match.finished = String.isNotEmpty(match.finished)
+		and match.finished ~= 'false' and match.finished ~= '0'
+		and 'true' or nil
 
 	if not noscore then
-		for scoreIndex, score in Table.iter.spairs(IndScore, StarcraftFfaInput.placementSortFunction) do
+		for scoreIndex, score in Table.iter.spairs(IndScore, StarcraftFfaInput._placementSortFunction) do
+			local opponent = match['opponent' .. scoreIndex]
 			counter = counter + 1
-			if counter == 1 and (match.winner or '') == '' then
+			if counter == 1 and String.isEmpty(match.winner) then
 				if match.finished or score >= match.bestof then
 					match.winner = scoreIndex
 					match.finished = 'true'
-					match['opponent' .. scoreIndex].placement = tonumber(match['opponent' .. scoreIndex].placement or '') or counter
-					match['opponent' .. scoreIndex].extradata.advances = true
-					match['opponent' .. scoreIndex].extradata.bg = match['opponent' .. scoreIndex].extradata.bg
-						or match.pbg[match['opponent' .. scoreIndex].placement] or 'down'
+					opponent.placement = tonumber(opponent.placement or '') or counter
+					opponent.extradata.advances = true
+					opponent.extradata.bg = String.nilIfEmpty(opponent.extradata.bg)
+						or match.pbg[opponent.placement]
+						or 'down'
 					temp.place = counter
 					temp.score = IndScore[scoreIndex]
 				else
 					break
 				end
 			elseif match.finished then
 				if temp.score == score then
-					match['opponent' .. scoreIndex].placement = tonumber(match['opponent' .. scoreIndex].placement or '') or temp.place
+					opponent.placement = tonumber(opponent.placement or '') or temp.place
 				else
-					match['opponent' .. scoreIndex].placement = tonumber(match['opponent' .. scoreIndex].placement or '') or counter
+					opponent.placement = tonumber(opponent.placement or '') or counter
 					temp.place = counter
 					temp.score = IndScore[scoreIndex]
 				end
-				match['opponent' .. scoreIndex].extradata.bg = match['opponent' .. scoreIndex].extradata.bg
-						or match.pbg[match['opponent' .. scoreIndex].placement] or 'down'
-				if match['opponent' .. scoreIndex].extradata.bg == 'up' then
-					match['opponent' .. scoreIndex].extradata.advances = true
+				opponent.extradata.bg = String.nilIfEmpty(opponent.extradata.bg)
+					or match.pbg[opponent.placement]
+					or 'down'
+				if opponent.extradata.bg == 'up' then
+					opponent.extradata.advances = true
 				end
 			else
 				break
 			end
 		end
 	elseif tonumber(match.winner or '') then
-		for oppIndex = 1, OppNumber do
-			match['opponent' .. oppIndex].placement = tonumber(match['opponent' .. oppIndex].placement or '') or 99
-			if match['opponent' .. oppIndex].placement == 99 and tonumber(match.winner) == oppIndex then
-				match['opponent' .. oppIndex].placement = 1
+		for oppIndex = 1, numberOfOpponents do
+			local opponent = match['opponent' .. oppIndex]
+			opponent.placement = tonumber(opponent.placement or '') or 99
```

Comment:
Is the `or 99` need? If so, to the top. If it can be worked away, great!
(same for a few other places that uses `99`)

Reply:
yeah it is needed for sorting later on
will make it a var